packages/libs/src/sdk/middleware/addRequestSignatureMiddleware.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  type Middleware,
+  type RequestContext,
+  type FetchParams
+} from '../api/generated/default'
+import { ServicesContainer } from '../types'
+
+const SIGNATURE_EXPIRY_MS = 60 * 1000
+const MESSAGE_HEADER = 'Encoded-Data-Message'
+const SIGNATURE_HEADER = 'Encoded-Data-Signature'
+
+let message: string | null = null
+let signature: string | null = null
+let timestamp: number | null = null
+
+/**
+ * Appends request authentication headers to a request.
+ * Request headers are computed only every SIGNATURE_EXPIRY_MS.
+ * @param options the middleware options
+ */
+export const addRequestSignatureMiddleware = ({
+  services
+}: {
+  services: Pick<ServicesContainer, 'auth' | 'logger'>
+}): Middleware => {
+  return {
+    pre: async (context: RequestContext): Promise<FetchParams> => {
+      const { auth, logger } = services
+      try {
+        const currentTimestamp = new Date().getTime()
+        const isExpired =
+          !timestamp || timestamp + SIGNATURE_EXPIRY_MS < currentTimestamp
```

Comment:
Thinking ahead, I don't remember our last discussion on this - are we planning to have signatures unique to the payload in the future? If so, we'd lose this caching perf win. I wonder if the impact would be too much if we went that route...

Also, if we aren't already, should we pad the expiry a bit to increase the odds that by the time it reaches DN it hasn't expired?

Reply:
there's no expiry on the DN side atm anyway so we're ok here!

Reply:
True, but maybe worth thinking about for when we do!

Reply:
yes definitely